File: android/app/src/main/java/com/martijnmichel/directusexpo/widget/directus/DirectusWidgetUrls.kt
package com.martijnmichel.directusexpo.widget.directus

import java.net.URL

object DirectusWidgetUrls {
  fun resolveInstanceBaseUrl(instanceUrl: String?, webhookUrl: String?): String? {
    fun normalize(s: String?): String? {
      if (s == null) return null
      var v = s.trim()
      if (v.isEmpty()) return null
      if (!v.startsWith("http://") && !v.startsWith("https://")) v = "https://$v"
      return v.replace(Regex("/+$"), "")
    }

    return normalize(instanceUrl)
      ?: run {
        val w = webhookUrl?.trim().orEmpty()
        if (w.isEmpty()) return null
        val url = try {
          URL(w)
        } catch (_: Exception) {
          return null
        }
        val scheme = url.protocol ?: "https"
        val host = url.host ?: return null
        val port = url.port
        if (port > 0) "$scheme://$host:$port" else "$scheme://$host"
      }
  }

  fun assetUrl(instanceBase: String, fileId: String, query: String): String =
    "${instanceBase.trimEnd('/')}/assets/$fileId?$query"
}

File: targets/widgets/DirectusWidgetUrls.swift
import Foundation

enum DirectusWidgetUrls {
  /// `https://host[:port]` from a flow trigger URL (matches Android `DirectusWidgetUrls.resolveInstanceBaseUrl` webhook branch).
  static func instanceBaseURLString(from webhookUrl: String?) -> String? {
    guard let webhookUrl, !webhookUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
          let url = URL(string: webhookUrl),
          let scheme = url.scheme,
          let host = url.host
    else {
      return nil
    }

    if let port = url.port {
      return "\(scheme)://\(host):\(port)"
    }
    return "\(scheme)://\(host)"
  }

  static func normalizedInstanceURLString(_ instanceUrl: String?) -> String? {
    guard var s = instanceUrl?.trimmingCharacters(in: .whitespacesAndNewlines), !s.isEmpty else {
      return nil
    }
    if !s.lowercased().hasPrefix("http://") && !s.lowercased().hasPrefix("https://") {
      s = "https://\(s)"
    }
    s = s.replacingOccurrences(of: "/+$", with: "", options: .regularExpression)
    return s
  }

  static func assetURLString(instanceBase: String, fileId: String, query: String) -> String {
    let base = instanceBase.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    return "\(base)/assets/\(fileId)?\(query)"
  }
}
